concurrency/worker/internal/heap/heap.go

```diff
@@ -0,0 +1,119 @@
+// Package heap provides heap operations for any type that implements
+// heap.Interface. A heap is a tree with the property that each node is the
+// minimum-valued node in its subtree.
+//
+// The minimum element in the tree is the root, at index 0.
+//
+// A heap is a common way to implement a priority queue. To build a priority
+// queue, implement the Heap interface with the (negative) priority as the
+// ordering for the Less method, so Push adds items while Pop removes the
+// highest-priority item from the queue. The Examples include such an
+// implementation; the file example_pq_test.go has the complete source.
+//
+// Note: This differes from the standard library in that it implements generics.
+package heap
+
+import (
+	"context"
+	"sort"
+)
+
+// The Interface type describes the requirements
+// for a type using the routines in this package.
+// Any type that implements it may be used as a
+// min-heap with the following invariants (established after
+// [Init] has been called or if the data is empty or sorted):
+//
+//	!h.Less(j, i) for 0 <= i < h.Len() and 2*i+1 <= j <= 2*i+2 and j < h.Len()
+//
+// Note that [Push] and [Pop] in this interface are for package heap's
+// implementation to call. To add and remove things from the heap,
+// use [heap.Push] and [heap.Pop].
+type Interface[T any] interface {
+	sort.Interface
+	Push(ctx context.Context, x T) // add x as element Len()
```

Comment:
Why bother passing `ctx` here? The stdlib impl doesn't as far as I can tell, and you don't actually use it in your impl either.

I guess it gives some more flexibility to the interface because you could smuggle logging details in on the ctx, and/or have some remote-backed heap implementation (where push/pop aren't in-memory but actually making network calls or disk calls or something where you might want to pass a context you can cancel?)

Reply:
At some point I'm going to add OTEL stuff into this.  Just haven't done it yet.